collector/collector.cpp

```diff
@@ -107,9 +107,14 @@ int InsertModule(int fd, const std::unordered_map<std::string, std::string>& arg
     args_str += entry.first + "=" + entry.second;
   }
   CLOG(DEBUG) << "Kernel module arguments: " << args_str;
-  int res = finit_module(fd, args_str.c_str(), 0);
-  if (res != 0) return res;
   struct stat st;
+  fstat(fd, &st);
+  size_t image_size = st.st_size;
+  void *image = malloc(image_size);
+  read(fd, image, image_size);
```

Comment:
We probably should check the return code, regarding `EINTR` to re-start reading like in:
````
static int __read(int fd, void *buf, int buflen)
{
        unsigned char *p = buf;
        int n, i = 0;

        while (i < buflen) {
                n = read(fd, p+i, buflen-i);
                if (n < 0) {
                        if (errno == EINTR)
                                continue;
                        else
                                return n;
                } else if (n == 0) {
                        return i;
                } else {
                        i += n;
                }
        }

        return i;
}
````

Reply:
thanks, added code for partial reads.